File: App/ore.cpp
/*
 Copyright (C) 2016 Quaternion Risk Management Ltd
 All rights reserved.

 This file is part of ORE, a free-software/open-source library
 for transparent pricing and risk analysis - http://opensourcerisk.org

 ORE is free software: you can redistribute it and/or modify it
 under the terms of the Modified BSD License.  You should have received a
 copy of the license along with this program.
 The license is also available online at <http://opensourcerisk.org>

 This program is distributed on the basis that it will form a useful
 contribution to risk analytics and model standardisation, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 FITNESS FOR A PARTICULAR PURPOSE. See the license for more details.
*/

#ifdef BOOST_MSVC
// disable warning C4503: '__LINE__Var': decorated name length exceeded, name was truncated
// This pragma statement needs to be at the top of the file - lower and it will not work:
// http://stackoverflow.com/questions/9673504/is-it-possible-to-disable-compiler-warning-c4503
// http://boost.2283326.n4.nabble.com/General-Warnings-and-pragmas-in-MSVC-td2587449.html
#pragma warning(disable : 4503)
#endif


#include <orea/app/oreapp.hpp>

#include <orea/app/initbuilders.hpp>

#include <qle/version.hpp>
#include <qle/gitversion.hpp>

#include <boost/program_options.hpp>

#include <iostream>

namespace po = boost::program_options;

#if !defined(BOOST_ALL_NO_LIB) && defined(BOOST_MSVC)
#include <orea/auto_link.hpp>
#include <ored/auto_link.hpp>
#include <ql/auto_link.hpp>
#include <qle/auto_link.hpp>
// Find the name of the correct boost library with which to link.
#define BOOST_LIB_NAME boost_serialization
#include <boost/config/auto_link.hpp>
#define BOOST_LIB_NAME boost_date_time
#include <boost/config/auto_link.hpp>
#define BOOST_LIB_NAME boost_system
#include <boost/config/auto_link.hpp>
#define BOOST_LIB_NAME boost_program_options
#include <boost/config/auto_link.hpp>
#define BOOST_LIB_NAME boost_timer
#include <boost/config/auto_link.hpp>
#define BOOST_LIB_NAME boost_chrono
#include <boost/config/auto_link.hpp>
#endif

using namespace std;
using namespace ore::data;
using namespace ore::analytics;

int main(int argc, char** argv) {

    if (argc == 2 && (string(argv[1]) == "-v" || string(argv[1]) == "--version")) {
        cout << "ORE version " << OPEN_SOURCE_RISK_VERSION << endl;
        exit(0);
    }

    if (argc == 2 && (string(argv[1]) == "-h" || string(argv[1]) == "--hash")) {
        #ifdef GIT_HASH
        cout << "Git hash " << GIT_HASH << endl;
        #endif
        exit(0);
    }

    // Parse command line arguments. The following form is supported:
    //   ore --input path/to/ore.xml [--output OutputDir]
    // In addition, any other --<name> <value> option is applied as an override
    // to the "setup" group of the parameters, e.g. --logMask 255, --marketDataFile market.txt.
    string inputFile;
    string outputPath;

    po::options_description desc("Allowed options");
    desc.add_options()
        ("help", "Produce help message")
        ("input,i", po::value<string>(&inputFile), "Input file (path to ore.xml)")
        ("output,o", po::value<string>(&outputPath), "Output directory, overrides setup/outputPath");

    po::variables_map vm;
    std::vector<std::string> unrecognized;
    try {
        po::parsed_options parsed = po::command_line_parser(argc, argv)
                                        .options(desc)
                                        .allow_unregistered()
                                        .run();
        po::store(parsed, vm);
        po::notify(vm);
        unrecognized = po::collect_unrecognized(parsed.options, po::include_positional);
    } catch (const std::exception& e) {
        cout << "error parsing command line: " << e.what() << endl << endl << desc << endl;
        return -1;
    }

    if (vm.count("help")) {
        cout << "usage: ORE path/to/ore.xml" << endl
             << "       ORE --input path/to/ore.xml [--output OutputDir] [--<setupParam> <value> ...]" << endl
             << endl
             << desc << endl;
        return 0;
    }

    std::map<std::string, std::string> setupOverrides;
    for (std::size_t i = 0; i < unrecognized.size(); ++i) {
        const std::string& token = unrecognized[i];
        if (token.size() > 2 && token.substr(0, 2) == "--") {
            std::string key = token.substr(2);
            std::string value;
            // Support both "--name value" and "--name=value".
            auto eq = key.find('=');
            if (eq != std::string::npos) {
                value = key.substr(eq + 1);
                key = key.substr(0, eq);
            } else if (i + 1 < unrecognized.size() && unrecognized[i + 1].substr(0, 2) != "--") {
                value = unrecognized[++i];
            }
            if (key.empty()) {
                cout << "error parsing command line: empty option name" << endl << endl << desc << endl;
                return -1;
            }
            setupOverrides[key] = value;
        } else if (inputFile.empty()) {
            // A single bare token is treated as the input file (backward compatible positional form).
            inputFile = token;
        } else {
            cout << "error parsing command line: unexpected argument '" << token << "'" << endl << endl
                 << desc << endl;
            return -1;
        }
    }

    if (inputFile.empty()) {
        cout << endl
             << "usage: ORE path/to/ore.xml" << endl
             << "       ORE --input path/to/ore.xml [--output OutputDir] [--<setupParam> <value> ...]" << endl
             << endl
             << desc << endl;
        return -1;
    }

    ore::analytics::initBuilders();

    try {
        auto params = QuantLib::ext::make_shared<Parameters>();
        params->fromFile(inputFile);
        if (!outputPath.empty())
            params->set("setup", "outputPath", outputPath);
        for (const auto& kv : setupOverrides) {
            cout << "overriding setup/" << kv.first << " = " << kv.second << endl;
            params->set("setup", kv.first, kv.second);
        }
        OREApp ore(params, true);
        ore.run();
        return 0;
    } catch (const exception& e) {
        cout << endl << "an error occurred: " << e.what() << endl;
        return -1;
    }
}

File: Examples/ORE-Python/ore.py
'''
 Copyright (C) 2018-2023 Quaternion Risk Management Ltd

 This file is part of ORE, a free-software/open-source library
 for transparent pricing and risk analysis - http://opensourcerisk.org
 ORE is free software: you can redistribute it and/or modify it
 under the terms of the Modified BSD License.  You should have received a
 copy of the license along with this program.
 The license is also available online at <http://opensourcerisk.org>
 This program is distributed on the basis that it will form a useful
 contribution to risk analytics and model standardisation, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 FITNESS FOR A PARTICULAR PURPOSE. See the license for more details.
'''

import sys, time
from ORE import *

#############################################
# Read inputs from files and kick off ORE run

print ("Loading parameters...")
params = Parameters()
params.fromFile("Input/ore.xml")

print ("Creating OREApp...")
ore = OREApp(params, True)

print ("Running ORE process...");
ore.run()

print ("Running ORE process done");

###########################################
# Check the analytics we have requested/run

print ("\nRequested analytics:")
analyticTypes = ore.getAnalyticTypes()
for name in analyticTypes:
    print("analytc:", name)

print("\npress <enter> ...")
sys.stdin.readline()

#####################################
# List all results, reports and cubes

print ("Result reports:");
reportNames = ore.getReportNames()
for name in reportNames:
    print("report:", name)

print ("\nResult cubes:");
cubeNames = ore.getCubeNames()
for name in cubeNames:
    print("cube:", name)

print("\npress <enter> ...")
sys.stdin.readline()

#######################
# Access report details

# pick one
reportName = "exposure_nettingset_CPTY_A"
print ("Load report", reportName)
report = ore.getReport(reportName)

# see PlainInMemoryReport
columnTypes = { 0: "Size",
                1: "Real",
                2: "string",
                3: "Date",
                4: "Period" }

print ("columns:", report.columns())
print ("rows:", report.rows())
for i in range(0, report.columns()):
    print("colum", i, "header", report.header(i), "type", report.columnType(i), columnTypes[report.columnType(i)])

print("\npress <enter> ...")
sys.stdin.readline()

time = report.dataAsReal(2);
epe = report.dataAsReal(3);
ene = report.dataAsReal(4);
print ("#Time", "EPE")
for i in range(0, report.rows()):
    print("%5.2f  %12.2f  %12.2f" % (time[i], epe[i], ene[i]))

print("\npress <enter> ...")
sys.stdin.readline()

#####################
# Access cube details

cubeName = "cube"
print ("Load NPV cube:", cubeName)
cube = ore.getCube(cubeName)
print ("cube ids:", cube.numIds())
print ("cube samples:", cube.samples())
print ("cube dates:", cube.numDates())
print ("cube depth:", cube.depth())

print("\npress <enter> ...")
sys.stdin.readline()

cubeIds = cube.ids() 
cubeDates = cube.dates()

for i in range (0, cube.numIds()):
    for j in range (0, cube.numDates()):
        for k in range (0, cube.samples()):
            for d in range (0, cube.depth()):
                npv = cube.get(i, j, k, d)
                print ("%s %s %4d %d %10.2f" % (cubeIds[i], cubeDates[j].ISO(), k, d, npv))

###########################################################
# Access one analytic e.g. to query the market or portfolio

print("\npress <enter> ...")
sys.stdin.readline()

print("Inspect NPV analytic ...")
analytic = ore.getAnalytic("NPV")
market  = analytic.getMarket()

print("Market asof:", market.asofDate())

curve = market.discountCurve("EUR")
print("EUR discount factor at 10Y =", curve.discount(10 * Years))

portfolio = analytic.portfolio()
tradeIds = portfolio.ids();
print("Portfolio size:", portfolio.size())

for tid in tradeIds:
    trade = portfolio.get(tid)
    print("trade id=%s type=%s" % (trade.id(), trade.tradeType()))
    
print("Done")
